### PR TITLE
Message action sample widget

### PR DESCRIPTION
**Pull request by qiqingjin**

### CODE REVIEW

File: samples/widgets/subscribe/README.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+# Subscribe widget
+
+This sample demonstrates how to use message action in widget.
+
+## How to use the sample
+Clone the [sample repo](https://github.com/esri/arcgis-experience-builder-sdk-resources) and copy this widget's folder (within `samples/widgets`) to the `client/your-extensions/widgets` folder of your Experience Builder installation.
+
+## How it works
+  Within `query-action-setting.tsx`,create an class extends the React.PureComponent class with the types `ActionSettingProps`,
```
Comment (gavinr) [inline samples/widgets/subscribe/README.md:9]:
```suggestion
  Within `query-action-setting.tsx`, create a class extends the React.PureComponent class with the type `ActionSettingProps`.
```

File: samples/widgets/subscribe/README.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  the configuration of the message action can be released through the `onSettingChange` method
```
Comment (gavinr) [inline samples/widgets/subscribe/README.md:10]:
```suggestion
  The configuration of the message action can be released through the `onSettingChange` method.
```

File: samples/widgets/subscribe/README.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    dataViewId: this.props.config.useDataSource.dataViewId,
+    rootDataSourceId: this.props.config.useDataSource.rootDataSourceId,
+    fields: allSelectedFields.map(f => f.jimuName)
+  })
+});
+
+```
+
+Within `query-action.tsx`, create an class extends the `AbstractMessageAction` class, we should set the uri of the action setting component through the `getSettingComponentUri` method, the `filterMessageType` method can filter out actions that do not meet our requirements. We can monitor the trigger of
+the action in the `onExecute` method, and can perform the operation what we want
```
Comment (gavinr) [inline samples/widgets/subscribe/README.md:27]:
```suggestion
```

File: samples/widgets/subscribe/README.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    mainDataSourceId: this.props.config.useDataSource.mainDataSourceId,
+    dataViewId: this.props.config.useDataSource.dataViewId,
+    rootDataSourceId: this.props.config.useDataSource.rootDataSourceId,
+    fields: allSelectedFields.map(f => f.jimuName)
+  })
+});
+
+```
+
+Within `query-action.tsx`, create an class extends the `AbstractMessageAction` class, we should set the uri of the action setting component through the `getSettingComponentUri` method, the `filterMessageType` method can filter out actions that do not meet our requirements. We can monitor the trigger of
```
Comment (gavinr) [inline samples/widgets/subscribe/README.md:26]:
```suggestion
Within `query-action.tsx`, create a class extends the `AbstractMessageAction` class. Then set the URI of the action setting component through the `getSettingComponentUri` method. Use the `filterMessageType` method to filter out actions that are not relevant to this widget. The `onExecute` method is called when the action is triggered.
```

**Review by gavinr — APPROVED** [review]

**Comment by qlqllu** [comment]
Let's rename this widget to `message subscriber`

File: samples/widgets/subscribe/README.md (hunk 1 above)
Comment (qlqllu) [inline samples/widgets/subscribe/README.md:9]:
```suggestion
Within `query-action-setting.tsx`,create a class extends the React.PureComponent class with the types `ActionSettingProps`,
```

File: samples/widgets/subscribe/README.md (hunk 2 above)
Comment (qlqllu) [inline samples/widgets/subscribe/README.md:10]:
```suggestion
the configuration of the message action can be changed through the `onSettingChange` method
```

File: samples/widgets/subscribe/src/setting/setting.tsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
+import {React, Immutable, IMFieldSchema, UseDataSource, DataSource, DataSourceTypes} from 'jimu-core';
+import {BaseWidgetSetting} from 'jimu-for-builder';
+import {FieldSelector} from 'jimu-ui/advanced/data-source-selector';
+import {DataSourceSelector} from 'jimu-ui/advanced/data-source-selector';
+
+export default class Setting extends BaseWidgetSetting{
```
Comment (qlqllu) [inline samples/widgets/subscribe/src/setting/setting.tsx:6]: Please do not extend from `BaseWidgetSetting`

**Comment by qlqllu** [comment]
Please clean code, this PR has a lots of useless code.

**Comment by chenbolan** [comment]
The comment in this pr has been modified accordingly in pr https://github.com/Esri/arcgis-experience-builder-sdk-resources/pull/63

**Pull request closed without merging**